lib/connection.js

```diff
@@ -1137,28 +1160,151 @@ Connection.prototype.search = function(sosl, callback) {
  */
 Connection.prototype.describe =
 Connection.prototype.describeSObject = function(type, callback) {
-  var url = [ this._baseUrl(), "sobjects", type, "describe" ].join('/');
-  return this.request(url).thenCall(callback);
+  var name = type.type ? type.type : type;
+  var url = [ this._baseUrl(), "sobjects", name, "describe" ].join('/');
+  var headers = type.ifModifiedSince 
+    ? { 'If-Modified-Since': type.ifModifiedSince } 
+    : {};
+  return this.request({
+    method: 'GET',
+    url: url,
+    headers: headers
+  }).then(function (resp) {
+    if (resp === '') {
+      return Promise.resolve(undefined);
+    } else {
+      return Promise.resolve(resp);
+    }
+  }).thenCall(callback);
 };
 
+/**
+ * Result returned by batchDescribeSObjects call
+ *
+ * @typedef {Object[]} DescribeSObjectResult
+ */
+/**
+ * Parameter for describeSObject call
+ * 
+ * @typedef {Object} BatchDescribeSObjectOptions
+ */
+/**
+ * Synonym of Connection#batchDescribe()
+ *
+ * @method Connection#batchDescribeSObjects
+ * @param {BatchDescribeSObjectOptions} options - options for function
+ * @param {String[]} options.types - names of objects to fetch
+ * @param {Boolean} options.autofetch - whether to automatically fetch metadata for large numbers of 
+ *                         types (one batch request returns a maximum of 25 results); when true, will make 
+ *                         subsequent requests until all object metadata is fetched; when false (default), 
+ *                         will make one batch request for maximum of 25 results
+ * @param {Callback.<DescribeSObjectResult[]>} [callback] - Callback function
+ * @returns {Promise.<DescribeSObjectResult[]>}
+ */
+/**
+ * Batch describe SObject metadata
+ *
+ * @method Connection#batchDescribe
+ * @param {BatchDescribeSObjectOptions} options - options for function
+ * @param {String[]} options.types - names of objects to fetch
+ * @param {Boolean} options.autofetch - whether to automatically fetch metadata for large numbers of 
+ *                         types (one batch request returns a maximum of 25 results); when true, will make 
+ *                         subsequent requests until all object metadata is fetched; when false (default), 
+ *                         will make one batch request for maximum of 25 results
+ * @param {Callback.<DescribeSObjectResult[]>} [callback] - Callback function
+ * @returns {Promise.<DescribeSObjectResult[]>}
+ */
+Connection.prototype.batchDescribe = Connection.prototype.batchDescribeSObjects = function (
+  options,
+  callback
+) {
+  var self = this;
+  var types = options.types;
+  var autofetch = options.autofetch || false;
+  var batches = [];
+  do {
+    var batch = types.length > MAX_BATCH_REQUESTS ? types.slice(0, MAX_BATCH_REQUESTS) : types;
+    batches.push(batch);
+    types = types.length > MAX_BATCH_REQUESTS ? types.slice(MAX_BATCH_REQUESTS) : [];
+  } while (types.length > 0 && autofetch);
+  Promise.all(
```

Comment:
We still want to return this, don't we?

Also, ca we run into max concurrent [API request limits](https://developer.salesforce.com/docs/atlas.en-us.salesforce_app_limits_cheatsheet.meta/salesforce_app_limits_cheatsheet/salesforce_app_limits_platform_api.htm)? I guess it would have to be over 625 objects? That doesn't seem that crazy to hit that limit.

Reply:
We do want to return it, yes. I made the change in `fix(conn): return the promise`. Certainly there are orgs with over 625 objects. Should I limit the concurrent requests to batches of 25 (or fewer)?

Reply:
Probably. Can we make it configurable? Maybe in batches of 15 by default. I'm choosing 15 because there could be other tools and services making API calls. I'm suggesting configurable to account for DE and trail orgs. It would probably be pretty rare for a trial org to have over 625 objects but I bet it would happen eventually.

Reply:
Ok I gave it a try. The option `maxConcurrentRequests` has a default and max of 15. The logic is getting pretty complex now. The function sets up the request batches (each length 15 or less) of batch requests(each length 25 or less). Then sequentially (via recursion in promise chain) sets off a batches of concurrent (via Promise.all()) requests.

Reply:
@jgellin-sf 🎉 Awesome that you are taking care of parallelization in this PR.

@amphro Worth noting that max concurrent API request limits only applies to long running requests:
> concurrent requests (calls) **with a duration of 20 seconds or longer**.

So I suppose it would be a rare case and if we handle the error properly it should be okay. I think the actual fail-safe way to handle it is to wrap each parallelized promise with a retry block so when it's blocked by the limit it's guaranteed to retry and fetch the results when the other long running requests finish, but it might not be necessary.

Reply:
That's interesting. I didn't even notice that. I thought I ran into this before with short-lived calls before but I just ran some test and sure enough. 

I think having each one in a retry logic is a little overkill, but given the way concurrent request work maybe this is overkill too. @jgellin-sf, I know you already did the work =/ but do you still think it is worth it?